pgjdbc/src/main/java/org/postgresql/copy/PGCopyInputStream.java

```diff
@@ -85,27 +88,32 @@ public int read(byte[] buf) throws IOException {
 
   public int read(byte[] buf, int off, int siz) throws IOException {
     checkClosed();
-    int got = 0;
+    int bytesRead = 0;
     boolean didReadSomething = false;
-    while (got < siz && (didReadSomething = gotBuf())) {
-      buf[off + got++] = this.buf[at++];
+    while (bytesRead < siz && (didReadSomething = isBufAvailable())) {
+      buf[off + bytesRead++] = this.buf[curPosition++];
```

Comment:
Why is this being copied a single byte at time instead of use arrayCopy?

Reply:
Well looking at this again; because ```isBufAvailable()``` can change buf!

Reply:
So certainly needs to be a loop, but should be able to perform bulk copy operations.

Reply:
I think we want something like this:

```
  public int read(byte[] buf, int off, int siz) throws IOException {
    checkClosed();
    if (siz == 0) {
      return 0;
    }
    int bytesRead = 0;
    while (bytesRead < siz && isBufAvailable()) {
        //we want to copy lesser of remaining requested or available in the buffer
        final int toCopy = Math.min(siz - bytesRead, len - curPosition);
        System.arrayCopy(this.buf, curPosition, buf, off + bytesRead, toCopy);
        bytesRead += toCopy;
        curPosition += toCopy;
    }
    //since we check size first, bytesRead being 0 means we have reached end of stream
    return bytesRead > 0 ? bytesRead : -1;
  }
```